scripts/upgrade-facets.js

```diff
@@ -0,0 +1,332 @@
+const { deployProtocolHandlerFacets } = require("./util/deploy-protocol-handler-facets.js");
+const { FacetCutAction, getSelectors, removeSelectors } = require("./util/diamond-utils.js");
+const { deploymentComplete, getFees, readContracts, writeContracts } = require("./util/utils.js");
+const { getInterfaceIds, interfaceImplementers } = require("./config/supported-interfaces.js");
+const Role = require("./domain/Role");
+const packageFile = require("../package.json");
+const readline = require("readline");
+const rl = readline.createInterface({
+  input: process.stdin,
+  output: process.stdout,
+});
+
+/**
+ * Upgrades or removes existing facets, or adds new facets.
+ *
+ * Prerequisite:
+ * - Admin must have UPGRADER role. Use `manage-roles.js` to grant it.
+ *
+ * Process:
+ *  1.  Edit scripts/config/facet-upgrade.js.
+ *  1a. Provide a list of facets that needs to be upgraded (field "addOrUpgrade") or removed completely (field "remove")
+ *  1b. Optionally you can specify which selectors should be ignored (field "skip"). You don't have to specify "initialize()" since it's ignored by default
+ *  2. Update protocol version in package.json. If not, script will prompt you to confirm that version remains unchanged.
+ *  2. Run the appropriate npm script in package.json to upgrde facets for a given network
+ *  3. Save changes to the repo as a record of what was upgraded
+ */
+async function main() {
+  // Bail now if hardhat network
+  if (network === "hardhat") process.exit();
+
+  const chainId = (await ethers.provider.getNetwork()).chainId;
+  const contractsFile = readContracts(chainId, network);
+  let contracts = contractsFile.contracts;
+  const interfaceIds = await getInterfaceIds();
+  const interfaceIdFromFacetName = (facetName) => interfaceIds[interfaceImplementers[facetName]];
+
+  const divider = "-".repeat(80);
+  console.log(`${divider}\nBoson Protocol Contract Suite Upgrader\n${divider}`);
+  console.log(`⛓  Network: ${network}\n📅 ${new Date()}`);
+
+  // Check that package.json version was updated
+  if (packageFile.version == contractsFile.protocolVersion) {
+    const answer = await new Promise((resolve) => {
+      rl.question("Protocol version has not been updated. Proceed anyway? (y/n) ", resolve);
+    });
+    switch (answer.toLowerCase()) {
+      case "y":
+      case "yes":
+        break;
+      case "n":
+      case "no":
+      default:
+        process.exit(1);
+    }
+  }
+
+  // Get the accounts
+  const adminAddress = environments[network].adminAddress;
+
+  // If admin address is unspecified, exit the process
+  if (adminAddress == ethers.constants.AddressZero || !adminAddress) {
+    console.log("Admin address must not be zero address");
+    process.exit(1);
+  }
+  // Get signer for admin address
+  const adminSigner = await ethers.getSigner(adminAddress);
+
+  console.log("🔱 Admin account: ", adminAddress ? adminAddress : "not found" && process.exit());
```

Comment:
not a big deal, but the condition falsy on this line will never be achieved because of lines 71-73.